libs/aws/langchain_aws/agents/utils.py

```diff
@@ -0,0 +1,289 @@
+        parameters = invocation_input.get("parameters", [])
+        parameters_json = {}
+        for parameter in parameters:
+            parameters_json[parameter.get("name")] = parameter.get("value", "")
+
+        tool = f"{action_group}::{function}"
+        if _DEFAULT_ACTION_GROUP_NAME in action_group:
+            tool = f"{function}"
+        return [
+            BedrockAgentAction(
+                tool=tool,
+                tool_input=parameters_json,
+                log=response_text,
+                session_id=session_id,
+                trace_log=trace_log,
+            )
+        ]
+    except Exception as ex:
+        raise Exception("Parse exception encountered {}".format(repr(ex)))
+
+
+def _create_bedrock_agent(
+    bedrock_client: Any,
+    agent_name: str,
+    agent_resource_role_arn: str,
+    instruction: str,
+    foundation_model: str,
+    client_token: Optional[str] = None,
+    customer_encryption_key_arn: Optional[str] = None,
+    description: Optional[str] = None,
+    guardrail_configuration: Optional[GuardrailConfiguration] = None,
+    idle_session_ttl_in_seconds: Optional[int] = None,
+) -> Union[str, None]:
+    """
+    Creates the bedrock agent
+    """
+    create_agent_request: dict = {
+        "agentName": agent_name,
+        "agentResourceRoleArn": agent_resource_role_arn,
+        "foundationModel": foundation_model,
+        "instruction": instruction,
+    }
+
+    if description:
+        create_agent_request["description"] = description
+
+    if client_token:
+        create_agent_request["clientToken"] = client_token
+
+    if customer_encryption_key_arn:
+        create_agent_request["customerEncryptionKeyArn"] = customer_encryption_key_arn
+
+    if guardrail_configuration is not None:
+        create_agent_request["guardrailConfiguration"] = {
+            "guardrailIdentifier": guardrail_configuration["guardrail_identifier"],
+            "guardrailVersion": guardrail_configuration["guardrail_version"] or "DRAFT",
+        }
+
+    if idle_session_ttl_in_seconds:
+        create_agent_request["idleSessionTTLInSeconds"] = idle_session_ttl_in_seconds
+
+    create_agent_response = bedrock_client.create_agent(**create_agent_request)
+    request_id = create_agent_response.get("ResponseMetadata", {}).get("RequestId", "")
+    logger.info(f"Create bedrock agent call successful with request id: {request_id}")
+    agent_id = create_agent_response["agent"]["agentId"]
+    create_agent_start_time = time.time()
+    while time.time() - create_agent_start_time < 10:
+        agent_creation_status = (
+            bedrock_client.get_agent(agentId=agent_id)
+            .get("agent", {})
+            .get("agentStatus", {})
+        )
+        if agent_creation_status == "NOT_PREPARED":
+            return agent_id
+        else:
+            time.sleep(2)
+
+    logger.error(f"Failed to create bedrock agent {agent_id}")
+    raise Exception(f"Failed to create bedrock agent {agent_id}")
```

Comment:
Should we use TimeoutError instead of Exception here? Would there be other reasons agent creation can fail here, for example if the role is not correct, does the API provide info for those failures?

Reply:
Good call. Added broad exception handling to `create_agent` call and updated this to TimeoutError.